socs/agents/hwp_supervisor/agent.py

```diff
@@ -0,0 +1,773 @@
+        if timeout is -1:
+            timeout = self.default_timeout
+        op.start(**kwargs)
+        status, msg, session = op.wait(timeout=timeout)
+
+    def update(self, clients: HWPClients, hwp_state: HWPState):
+        """
+        Runs control operations and updates the control-state based on the
+        ``hwp_state`` object.
+
+        Args
+        --------
+        clients : HWPClients
+            Clients to use to control the HWP.
+        hwp_state : HWPState
+            Current state of the HWP.
+        """
+        if self.state_type == ControlStateType.NONE:
+            return
+
+        elif self.state_type == ControlStateType.PID_FREQ:
+            # Set the PID and then switch to WAIT_FOR_TARGET_FREQ
+            self.run_and_validate(clients.pid.set_direction,
+                                  kwargs={'direction': self.pid_dir})
+            self.run_and_validate(clients.pid.declare_freq,
+                                  kwargs={'freq': self.target_freq})
+            self.run_and_validate(clients.pid.use_ext)
+            self.run_and_validate(clients.pid.set_on)
+            self.set_state_type(ControlStateType.WAIT_FOR_TARGET_FREQ)
+
+        elif self.state_type == ControlStateType.PMX_OFF:
+            # Turn PMX off and then switch to WAIT_FOR_TARGET_FREQ
+            self.run_and_validate(clients.pid.set_off)
+            self.set_state_type(ControlStateType.WAIT_FOR_TARGET_FREQ)
+
+        elif self.state_type == ControlStateType.WAIT_FOR_TARGET_FREQ:
+            # Check if we are close enough to the target frequency.
+            # This will make sure we remain within the frequency threshold for
+            # ``self.freq_thresh_duration`` seconds before switching to DONE
+            f = hwp_state.pid_current_freq
+            if f is not None:
+                # If within_thresh for freq_thresh_duration, switch to Done
+                if np.abs(f - self.target_freq) < self.freq_thresh:
+                    if self.freq_within_thresh_start is None:
+                        self.freq_within_thresh_start = time.time()
+                    elif time.time() - self.freq_within_thresh_start > self.freq_thresh_duration:
+                        self.set_state_type(ControlStateType.DONE, success=True)
+                else:
+                    self.freq_within_thresh_start = None
+            else:
+                self.freq_within_thresh_start = None
+
+        elif self.state_type == ControlStateType.CONST_VOLT:
+            # Set to constant voltage mode
+            clients.pmx.ign_ext()
+            self.run_and_validate(clients.pmx.ign_ext)
+            self.run_and_validate(clients.pmx.set_v,
+                                  kwargs={'voltage': self.voltage})
+            self.set_state_type(ControlStateType.DONE, success=True)
+
+        elif self.state_type == ControlStateType.BRAKE:
+            # Actively set the PID to brake the HWP
+            self.run_and_validate(clients.pid.tune_stop)
```

Comment:
The `clients.pid.tune_stop` function reverses the rotation direction, changes the pid parameters, and set the setpoint to 0 Hz. If this function is run unattended, the HWP will reach 0 Hz and then start accelerating in the opposite direction. To fully stop the HWP the power to the PMX needs to be turned off as the rotation frequency approaches 0 Hz.